### PR TITLE
feat(attention): add variant_owns_mask for JIT variants that own the full mask

### PR DESCRIPTION
**Pull request by lesj0610**
## 📌 Description

A custom JIT attention variant can compute the complete attention mask inside its `REGISTER_LOGITS_MASK` hook from compact side metadata (for example per-query ranges passed as additional tensors), without ever materializing a `custom_mask`/`packed_custom_mask` tensor.

The FA2 batch prefill kernel only evaluates `LogitsMask` on **every** KV tile under `MaskMode::kCustom`; under `kNone`/`kCausal` it is evaluated only on the causal/window boundary tiles:

```cpp
// include/flashinfer/attention/prefill.cuh
if (MASK_MODE == MaskMode::kCustom) {
  logits_mask<KTraits>(...);
} else {
  if (iter >= mask_iteration || iter < window_iteration) {
    logits_mask<KTraits>(...);
  }
}
```

Today the only way to select `MaskMode::kCustom` is to pass a mask tensor to `plan()`. For a mask-owning variant this defeats the point: the dense mask costs `O(qo_len * kv_len)` memory even though the kernel never needs its contents (e.g. 16384 query tokens against a 262144-token KV is a 4 GiB boolean mask, still 512 MiB packed), and a variant relying on `causal=False` alone silently loses its mask on interior KV tiles once sequences span more than one KV tile.

This PR adds an explicit `variant_owns_mask: bool = False` constructor flag to `BatchPrefillWithPagedKVCacheWrapper` and `BatchPrefillWithRaggedKVCacheWrapper` that selects `MaskMode::kCustom` independently of the mask buffer. The flag is fail-closed on three fronts:

- requires `jit_args` — the default attention variant dereferences the custom mask buffer under `MaskMode::kCustom`;
- rejected for any backend other than `fa2`, before the JIT module is built — the SM90 batch prefill kernels return `cudaErrorNotSupported` under `MaskMode::kCustom` (`include/flashinfer/attention/hopper/prefill_sm90.cuh`);
- `plan()` rejects `prefix_len_ptr` — multi-item scoring selects `MaskMode::MULTIITEMSCORING`, which would silently override `CUSTOM`.

No existing code path changes behavior when the flag is left at its default.

## 🔍 Related Issues

N/A

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.) — I ran the targeted tests below plus the neighboring JIT-wrapper regression test on SM80, not the full suite.

`tests/utils/test_jit_example.py::test_batch_prefill_variant_owns_mask` uses a variant that owns a sliding-window mask on sequences long enough to contain interior KV tiles for every `CTA_TILE_KV` configuration, and asserts on SM80 FA2:

- with `variant_owns_mask=True`, both the ragged and the paged wrapper match a dense reference (`torch.testing.assert_close`);
- the same variant **without** the flag (`MaskMode::kNone`) does not apply the mask on interior tiles, pinning the boundary-only behavior this flag exists to bypass.

`test_variant_owns_mask_requires_jit_module` checks the constructor rejections (no JIT module; non-fa2 backend, raised before any JIT build), and `test_variant_owns_mask_rejects_multi_item_scoring` checks the `plan()` rejection of `prefix_len_ptr`.

```bash
pytest tests/utils/test_jit_example.py::test_batch_prefill_variant_owns_mask \
       tests/utils/test_jit_example.py::test_variant_owns_mask_requires_jit_module \
       tests/utils/test_jit_example.py::test_variant_owns_mask_rejects_multi_item_scoring
# 4 passed (the multi-item-scoring rejection is parameterized over both wrappers)
pytest tests/utils/test_jit_example.py::test_batch_prefill_flash_sigmoid
# 1 passed (JIT wrapper path regression check)
```

## Reviewer Notes

The run-time change is a single condition in each wrapper's `run()`: `if self._custom_mask_buf is not None or self._variant_owns_mask:`. `plan()`'s module selection is untouched — its `use_custom_mask` argument only feeds `determine_attention_backend()`, which is not reached when a JIT module is supplied.

## Summary by CodeRabbit

* **New Features**
  * Added an option for custom FA2 attention variants to provide and manage their own masks during batch prefill.
  * Supported both paged and ragged KV-cache workflows.
  * Added validation for required JIT configuration and unsupported scoring or backend combinations.

* **Tests**
  * Added coverage for custom masking behavior, reference output matching, hardware compatibility, and invalid configuration handling.

### CODE REVIEW

File: tests/utils/test_jit_example.py [hunk 1]
```diff
@@ -452,6 +452,252 @@ def test_batch_prefill_flash_sigmoid():
...
+        num_qo_heads,
+        num_kv_heads,
+        head_dim,
+        1,
+        causal=False,
+        q_data_type=torch.float16,
+        kv_data_type=torch.float16,
+    )
+    o_paged = wrapper_paged.run(q, (k, v), mask_window, sm_scale)
+    torch.testing.assert_close(o_paged.float(), o_ref, rtol=2e-2, atol=2e-2)
```
[the remark this answers is not in the document]
Comment (lesj0610) [reply tests/utils/test_jit_example.py:619]: Added in b36adf2

File: flashinfer/prefill.py [hunk 2]
```diff
@@ -1652,6 +1652,7 @@ def __init__(
         backend: str = "auto",
         jit_args: Optional[List[Any]] = None,
         jit_kwargs: Optional[Dict[str, Any]] = None,
+        variant_owns_mask: bool = False,
```
[the remark this answers is not in the document]
Comment (lesj0610) [reply flashinfer/prefill.py:1655]: Pre-existing pattern, leaving as is

**Comment by lesj0610** [comment]
@qsang-nv, could I ask you to take a look at this PR as well?

It adds an opt-in `variant_owns_mask` API for FA2 JIT batch-prefill variants, allowing them to select `MaskMode::kCustom` and compute the full mask from compact side metadata without allocating a dense or packed mask.

Your reviews on my previous FlashInfer PRs have caught several subtle wrapper/kernel contract issues, so I’d particularly value your review of the API and mask-mode selection contract here. The targeted SM80 tests and pre-commit checks pass; the full test suite was not run. Thank you.

**Review by qsang-nv — APPROVED** [review]
